plugins/hls-class-plugin/src/Ide/Plugin/Class.hs

```diff
@@ -63,25 +69,69 @@ addMethodPlaceholders state AddMinimalMethodsParams{..} = do
   medit <- liftIO $ runMaybeT $ do
     docPath <- MaybeT . pure . uriToNormalizedFilePath $ toNormalizedUri uri
     pm <- MaybeT . runAction "classplugin" state $ use GetParsedModule docPath
-    let
-      ps = pm_parsed_source pm
-      anns = relativiseApiAnns ps (pm_annotations pm)
-      old = T.pack $ exactPrint ps anns
-
     (hsc_dflags . hscEnv -> df) <- MaybeT . runAction "classplugin" state $ use GhcSessionDeps docPath
-    List (unzip -> (mAnns, mDecls)) <- MaybeT . pure $ traverse (makeMethodDecl df) methodGroup
-    let
-      (ps', (anns', _), _) = runTransform (mergeAnns (mergeAnnList mAnns) anns) (addMethodDecls ps mDecls)
-      new = T.pack $ exactPrint ps' anns'
-
+    (old, new) <- makeEditText pm df
     pure (workspaceEdit caps old new)
+
   forM_ medit $ \edit ->
     sendRequest SWorkspaceApplyEdit (ApplyWorkspaceEditParams Nothing edit) (\_ -> pure ())
   pure (Right Null)
   where
-
     indent = 2
 
+    workspaceEdit caps old new
+      = diffText caps (uri, old) new IncludeDeletions
+
+    toMethodName n
+      | Just (h, _) <- T.uncons n
+      , not (isAlpha h || h == '_')
+      = "(" <> n <> ")"
+      | otherwise
+      = n
+
+#if MIN_VERSION_ghc(9,2,0)
+    makeEditText pm df = do
+      List mDecls <- MaybeT . pure $ traverse (makeMethodDecl df) methodGroup
+      let ps = makeDeltaAst $ pm_parsed_source pm
+          old = T.pack $ exactPrint ps
+          (ps', _, _) = runTransform (addMethodDecls ps mDecls)
+          new = T.pack $ exactPrint ps'
+      pure (old, new)
+
+    makeMethodDecl df mName =
+        rightToMaybe . parseDecl df (T.unpack mName) . T.unpack
+            $ toMethodName mName <> " = _"
+
+    addMethodDecls ps mDecls = do
+      allDecls <- hsDecls ps
+      let (before, ((L l inst): after)) = break (containRange range . getLoc) allDecls
+      replaceDecls ps (before ++ (L l (addWhere inst)): (map newLine mDecls ++ after))
+      where
+        addWhere (InstD xInstD (ClsInstD ext decl@ClsInstDecl{..})) =
+          let ((EpAnn entry anns comments), key) = cid_ext
+          in InstD xInstD (ClsInstD ext decl {
+            cid_ext = (EpAnn
+                        entry
+                        (AddEpAnn AnnWhere (EpaDelta (SameLine 1) []) : anns)
+                        comments
+                      , key)
+          })
+        addWhere decl = decl
```

Comment:
can you add a brief comment explaining what this code does? It's kinda hard to read and therefore maintain

Reply:
Explained the effect of the function and the trace of the parameter.